File: detect_cycle2.py
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this :
#We use slow and fast pointers. When fast=slow that means there is a cycle
#Now then we make slow as head and increment both +1 again.
# Now when again they will meet then we return any of the pointer
# we run the initial loop and keep a flag , if flag is false then its no cycle
# If fast or fast.next is null then our loop exits.


# // Your code here along with comments explaining your approach

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow,fast=head,head
        flag=False
        
        if head==None:
            return None
        
        while fast!=None and fast.next!=None:
            slow=slow.next
            fast=fast.next.next
            if slow==fast:
                slow=head
                flag=True
                break
        
        if flag!=True:
            return None
        if flag:
            while fast!=slow:
                slow=slow.next
                fast=fast.next
                
        
        return slow

File: Detect Cycle - 2.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

//Time Complexity = O(2N);
//Space Complexity = O(1);
//Method: write mathematical equation for distance travelled by slow pointer and fast pointer, inorder to reach the cycle head, it takes equivalent distance from slow and head pointer.
public class Solution {
    public ListNode detectCycle(ListNode head) {
        //base
        if(head == null)return null;
        ListNode fast = head;
        ListNode slow = head;
        ListNode slower = head;

        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;

            if(slow == fast) break;
        }
        if(fast == null || fast.next == null) return null;


        while(slower != slow){
            slow = slow.next;
            slower = slower.next;
        }
        return slow;
    }
}